### PR TITLE
Handle procedural filtering in iOS

### PR DESCRIPTION
**Pull request by cuba**
Resolves

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js [hunk 1]
```diff
@@ -1070,19 +1084,96 @@ window.__firefox__.execute(function($) {
...
+    if (CC.proceduralActionFilters === undefined) {
+      return
+    }
+
+    /**
+    * Adds given styles to allStyleRules
+    * @param {Object} entry The entry to add
+    */
+    const addStyleRule = (entry) => {
+      const rule = entry.selector + '{' + entry.rules.join(';') + ';}'
```
Comment (antonok-edm) [inline ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js:1117]: `entry` is always 1 element long; no need to accept an array here

File: ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js [hunk 2]
```diff
@@ -1070,19 +1084,96 @@ window.__firefox__.execute(function($) {
...
+        // Check if the element has any classes to remove because
+        // classList.remove(tokens...) always triggers another mutation
+        // even if nothing was removed.
+        if (element.classList.contains(action.arg)) {
+          element.classList.remove(action.arg);
+        }
+      }
+    }
+    for (const { selector, action } of CC.proceduralActionFilters) {
+      console.debug('Processing procedural selector', selector, action);
```
Comment (antonok-edm) [inline ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js:1142]: probably don't want to spam the console with these; there could be a lot on e.g. sites with infinite scrolling feeds

File: ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js [hunk 3]
```diff
@@ -84,7 +90,14 @@ window.__firefox__.execute(function($) {
...
     urlFirstParty: new Map(),
-    alreadyKnownFirstPartySubtrees: new WeakSet()
+    alreadyKnownFirstPartySubtrees: new WeakSet(),
+    // All the procedural rules that exist and need to be processed
+    // when the script is loaded and a new element is added
+    proceduralActionFilters: undefined,
+    // Tells us if procedural filtering is available
+    hasProceduralActions: false,
+    // The map of style attributes and their values used for procedural filtering
+    styleAttrMap: new Map()
```
Comment (antonok-edm) [inline ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js:100]: appears to be unused?

**Review by antonok-edm — APPROVED** [review]

File: ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js (hunk 3 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js:100]: ahh yes remained from converting the desktop script

File: ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js (hunk 1 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Sandboxed/SelectorsPollerScript.js:1117]: i don't know if I updated the code (as this used to be originally an array in the past) but entry isn't an array here so im not sure what if anything I'm supposed to do here.

**Pull request merged**